src/dispatch/__init__.py

```diff
@@ -2,27 +2,97 @@
 
 from __future__ import annotations
 
+import os
+from concurrent import futures
+from http.server import ThreadingHTTPServer
+from typing import Any, Callable, Coroutine, Optional, TypeVar, overload
+from urllib.parse import urlsplit
+
+from typing_extensions import ParamSpec, TypeAlias
+
 import dispatch.integrations
 from dispatch.coroutine import all, any, call, gather, race
-from dispatch.function import DEFAULT_API_URL, Client, Registry, Reset
+from dispatch.function import DEFAULT_API_URL, Client, Function, Registry, Reset
+from dispatch.http import Dispatch
 from dispatch.id import DispatchID
 from dispatch.proto import Call, Error, Input, Output
 from dispatch.status import Status
 
 __all__ = [
+    "Call",
     "Client",
-    "DispatchID",
     "DEFAULT_API_URL",
+    "DispatchID",
+    "Error",
     "Input",
     "Output",
-    "Call",
-    "Error",
+    "Registry",
     "Reset",
     "Status",
-    "call",
-    "gather",
     "all",
     "any",
+    "call",
+    "function",
+    "gather",
     "race",
-    "Registry",
+    "run",
+    "serve",
 ]
+
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+_registry: Optional[Registry] = None
+
+
+def default_registry():
+    global _registry
+    if not _registry:
+        _registry = Registry()
+    return _registry
+
+
+@overload
+def function(func: Callable[P, Coroutine[Any, Any, T]]) -> Function[P, T]: ...
+
+
+@overload
+def function(func: Callable[P, T]) -> Function[P, T]: ...
+
+
+def function(func):
+    return default_registry().function(func)
+
+
+def run(init: Optional[Callable[P, None]] = None, *args: P.args, **kwargs: P.kwargs):
+    """Run the default dispatch server. The default server uses a function
+    registry where functions tagged by the `@dispatch.function` decorator are
+    registered.
+
+    This function is intended to be used with the `dispatch` CLI tool, which
+    automatically configures environment variables to connect the local server
+    to the Dispatch bridge API.
+
+    Args:
+        init: An initialization function called after binding the server address
+            but before entering the event loop to handle requests.
+
+        args: Positional arguments to pass to the entrypoint.
+
+        kwargs: Keyword arguments to pass to the entrypoint.
+
+    Returns:
+        The return value of the entrypoint function.
+    """
+    address = os.environ.get("DISPATCH_ENDPOINT_ADDR", "localhost:8000")
+    parsed_url = urlsplit("//" + address)
+    server_address = (parsed_url.hostname or "", parsed_url.port or 0)
+    server = ThreadingHTTPServer(server_address, Dispatch(default_registry()))
+    try:
+        if init is not None:
+            init(*args, **kwargs)
+        server.serve_forever()
+    finally:
+        server.shutdown()
+        server.server_close()
```

Comment:
This happens:

```
$ ./dispatch run -- ./main.py

  ╭───────────────────────────────────────────────────────────────────╮
  │                                                                   │
  │  Starting Dispatch session: ce77804d-b40d-4df7-b8a4-d10afcc4ab31  │
  │                                                                   │
  │  Run 'dispatch help run' to learn about Dispatch sessions.        │
  │                                                                   │
  ╰───────────────────────────────────────────────────────────────────╯

Hello, World!
127.0.0.1 - - [22/Apr/2024 18:22:23] "POST /dispatch.sdk.v1.FunctionService/Run HTTP/1.1" 200 -
^C
  ╭────────────────────────────────────────────────────────────────────────────────╮
  │                                                                                │
  │  To resume this Dispatch session:                                              │
  │                                                                                │
  │      dispatch run --session ce77804d-b40d-4df7-b8a4-d10afcc4ab31 -- ./main.py  │
  │                                                                                │
  ╰────────────────────────────────────────────────────────────────────────────────╯
```

Reply:
Ah right because we send `SIGTERM` rather than passing on the signal the CLI receives here: https://github.com/stealthrocket/dispatch/blob/00ad5441b8da130d378af1843af364241906b828/cli/run.go#L171. I think I had it passing on the signal originally, but `uvicorn` was handling `SIGINT` differently when it was run inside the CLI (and wasn't itself interactive).